File: plugins/probe-rat-layer/probe_layer/pipeline.py
"""pipeline：loader → 分层 → 后处理 → 结果 的编排。"""
from __future__ import annotations

import dataclasses
import math
from collections import defaultdict

from .config import LayeringConfig
from .model import (LayeringResult, LayerInfo, NetClass, ConflictLevel, Wire)
from .io.loader import LoadedData
from .cancel import LayeringCancelled, check_cancel
from .io import allegro_skill as askill
from .core import geometry as geo
from .core import congestion
from .core import conflict_classifier as cc
from .core import layer_stack as lstack
from .core import layer_packing as packing
from .core import optimizer as opt
from .core import metrics
from .core import graph_coloring as coloring
from .core import post_process as pp


def _usable_area(wires: tuple[Wire, ...], keepouts) -> float:
    if not wires:
        return 0.0
    xs = [p.x for w in wires for p in (w.start, w.end)]
    ys = [p.y for w in wires for p in (w.start, w.end)]
    area = (max(xs) - min(xs)) * (max(ys) - min(ys))
    for z in keepouts:
        if isinstance(z, geo.RectZone):
            area -= (z.xmax - z.xmin) * (z.ymax - z.ymin)
        else:
            area -= math.pi * z.radius * z.radius
    return max(area, 0.0)


def apply_feedback_penalties(data: LoadedData, unrouted: set[str],
                             amount: float = 1.0) -> LoadedData:
    """对未布通线累加 penalty（增量修复用）。返回新的 LoadedData。"""
    wires = tuple(
        Wire(w.wire_id, w.net_id, w.start, w.end, w.width, w.clearance,
             w.penalty + amount if w.wire_id in unrouted else w.penalty)
        for w in data.wires)
    return LoadedData(data.stack, data.signal_groups, data.net_groups, data.nets,
                      data.keepouts, wires, data.units, data.warnings)


def run_once(data: LoadedData, cfg: LayeringConfig,
             on_progress=None, cancel_event=None) -> LayeringResult:
    """单轮分层。`on_progress(percent, message)` 可选进度回调（0~100）；
    `cancel_event` 可选取消事件（is_set() 时抛 LayeringCancelled）。"""
    def emit(pct: float, msg: str) -> None:
        if on_progress:
            on_progress(pct, msg)

    emit(3, "分离电源/地")
    trace_nets, plane_nets = lstack.split_trace_plane(data.nets, cfg)
    trace_net_ids = {n.net_id for n in trace_nets}
    wires = tuple(w for w in data.wires if w.net_id in trace_net_ids)

    group_of = {n.net_id: (n.signal_group_id or "default") for n in trace_nets}
    wire_group_of = {w.wire_id: group_of.get(w.net_id, "default") for w in wires}
    allowed = {w.wire_id: lstack.wire_allowed_layers(w, trace_nets, data.signal_groups,
                                                     data.stack) for w in wires}
    pins = tuple(p for n in trace_nets for p in n.pins)

    check_cancel(cancel_event)
    emit(12, "构建拥塞图")
    cmap0 = congestion.build_congestion_map(wires, data.keepouts, pins, cfg)
    check_cancel(cancel_event)
    emit(22, "冲突检测")
    conflicts, hard_graph = cc.detect_all_conflicts(wires, data.keepouts, cfg, cmap0)
    check_cancel(cancel_event)

    usable = _usable_area(wires, data.keepouts)
    lb = packing.capacity_lower_bound(wires, wire_group_of, usable, cfg)

    warnings: list[str] = list(data.warnings)
    if cfg.method == "dsatur":
        try:
            assignment = coloring.dsatur_color(hard_graph, allowed=allowed)
        except coloring.UncolorableError as e:
            warnings.append(str(e))
            assignment = {}
    else:
        emit(32, "扇区轮询分层")
        assignment = packing.pack_layers(
            wires, allowed, data.keepouts, pins, cfg, hard_graph,
            on_progress=lambda f: emit(32 + f * 18, "扇区轮询分层"),
            cancel_event=cancel_event)
        # 软冲突（交叉）最小化 + 均衡精修
        soft_pairs = [(c.wire_a, c.wire_b) for c in conflicts
                      if c.level == ConflictLevel.SOFT]
        layers = sorted(set(assignment.values()))
        # 先跑一次快速贪心（系统化下降），作为精修的起点 / 对比基线
        if cfg.optimizer != "none":
            emit(52, "贪心交叉最小化")
            assignment = packing.minimize_crossings(
                assignment, soft_pairs, layers, hard_graph,
                max_passes=cfg.minimize_crossings_passes,
                cancel_event=cancel_event)
            check_cancel(cancel_event)
        # 再在贪心结果上做模拟退火，跳出贪心局部最优
        if cfg.optimizer == "sa":
            best = assignment
            best_soft = metrics.soft_crossings(best, soft_pairs)
            for r in range(max(1, cfg.sa_restarts)):
                c = dataclasses.replace(cfg, sa_seed=cfg.sa_seed + r)
                cand = opt.optimize_layering(
                    best, wires=wires, soft_pairs=soft_pairs,
                    hard_graph=hard_graph, allowed=allowed, cfg=c,
                    on_progress=lambda f: emit(55 + f * 33, "模拟退火精修"),
                    cancel_event=cancel_event)
                check_cancel(cancel_event)
                s = metrics.soft_crossings(cand, soft_pairs)
                if s < best_soft:
                    best, best_soft = cand, s
            assignment = best

    check_cancel(cancel_event)
    emit(90, "后处理与人工兜底")
    viol = pp.verify_hard_free(assignment, hard_graph)
    manual_nets: list[str] = []
    manual_wires: set[str] = set()
    if viol:
        # 少量分不开的线不强塞：从自动分层里移除，标记留给人工 route
        manual_wires = {w for pair in viol for w in pair}
        manual_nets = sorted({w.net_id for w in wires if w.wire_id in manual_wires})
        for w in manual_wires:
            assignment.pop(w, None)
        warnings.append(f"{len(manual_nets)} 条线需人工 route（同层硬冲突无法自动分层）: {manual_nets}")

    soft_per_layer = pp.soft_conflicts_per_layer(assignment, conflicts)
    detour, via = pp.collect_layer_marks(wires, assignment, conflicts, data.keepouts, cfg)
    occ_per_layer = pp.max_occupancy_per_layer(assignment, wires, data.keepouts, pins, cfg)

    net_by_id = {n.net_id: n for n in data.nets}
    by_layer: dict[int, list[Wire]] = defaultdict(list)
    for w in wires:
        if w.wire_id in assignment:
            by_layer[assignment[w.wire_id]].append(w)

    layer_infos: list[LayerInfo] = []
    for layer in sorted(by_layer):
        ws = by_layer[layer]
        nets = sorted({w.net_id for w in ws})
        groups = sorted({net_by_id[n].signal_group_id
                         for n in nets if n in net_by_id and net_by_id[n].signal_group_id})
        layer_infos.append(LayerInfo(
            layer_index=layer,
            kind=data.stack.kind_of(layer) if data.stack else "signal",
            signal_groups=tuple(groups),
            wires=tuple(sorted(w.wire_id for w in ws)),
            nets=tuple(nets),
            soft_conflict_count=soft_per_layer.get(layer, 0),
            max_occupancy=round(occ_per_layer.get(layer, 0.0), 4),
            requires_detour=tuple(sorted(detour.get(layer, set()))),
            requires_endpoint_via=tuple(sorted(via.get(layer, set()))),
        ))

    # plane 层（仅 net 列表）
    if plane_nets and cfg.plane_nets_excluded and data.stack is not None:
        for pl in data.stack.plane_layers():
            layer_infos.append(LayerInfo(layer_index=pl, kind="plane",
                                         nets=tuple(sorted(n.net_id for n in plane_nets))))

    layer_infos.sort(key=lambda li: li.layer_index)

    unassigned = [w.wire_id for w in wires
                  if w.wire_id not in assignment and w.wire_id not in manual_wires]
    if unassigned:
        warnings.append(f"以下线未能在允许层内分配: {unassigned}")

    hard = tuple(c for c in conflicts if c.level == ConflictLevel.HARD)
    soft = tuple(c for c in conflicts if c.level == ConflictLevel.SOFT)
    emit(100, "完成")
    return LayeringResult(
        layers=tuple(layer_infos),
        assignment=assignment,
        plane_nets=tuple(sorted(n.net_id for n in plane_nets)),
        hard_conflicts=hard,
        soft_conflicts=soft,
        method=cfg.method,
        iterations_used=1,
        capacity_lower_bound=lb,
        warnings=tuple(warnings),
        manual_route_nets=tuple(manual_nets),
    )


def run(data: LoadedData, cfg: LayeringConfig,
        feedback_path: str | None = None,
        on_progress=None, cancel_event=None) -> LayeringResult:
    """单轮分层；若提供 feedback 文件则迭代（占位，P3 真实闭环）。

    可选 `on_progress(percent, message)` / `cancel_event`（见 run_once）。
    """
    result = run_once(data, cfg, on_progress=on_progress, cancel_event=cancel_event)
    if not cfg.feedback_enabled or not feedback_path:
        return result

    iterations = 1
    for it in range(2, cfg.max_loop_iterations + 1):
        fb = askill.parse_route_feedback(feedback_path)
        if not fb.unrouted_all():
            break
        data = apply_feedback_penalties(data, fb.unrouted_all())
        result = run_once(data, cfg, on_progress=on_progress, cancel_event=cancel_event)
        iterations = it
        if not fb.unrouted_all():
            break
    # 更新迭代计数
    return LayeringResult(
        layers=result.layers, assignment=result.assignment,
        plane_nets=result.plane_nets, hard_conflicts=result.hard_conflicts,
        soft_conflicts=result.soft_conflicts, method=result.method,
        iterations_used=iterations, capacity_lower_bound=result.capacity_lower_bound,
        warnings=result.warnings, manual_route_nets=result.manual_route_nets,
    )

File: core-plugins/probe-rat-layer/src/pipeline.rs
//! pipeline：loader → 分层 → 后处理 → 结果 的编排（移植自 Python `pipeline.py`）。

use crate::cancel::LayeringCancelled;
use crate::config::LayeringConfig;
use crate::layer_packing;
use crate::io::LoadedData;
use crate::model::{
    Conflict, ConflictLevel, KeepoutZone, LayerInfo, LayeringResult, Net, Pin, Wire,
};
use crate::optimizer;
use crate::state::Progress;
use crate::{
    conflict_classifier as cc, congestion, graph_coloring as coloring, layer_stack as lstack,
    metrics, post_process as pp,
};
use std::collections::{HashMap, HashSet};

fn _usable_area(wires: &[Wire], keepouts: &[KeepoutZone]) -> f64 {
    if wires.is_empty() {
        return 0.0;
    }
    let mut xs: Vec<f64> = Vec::new();
    let mut ys: Vec<f64> = Vec::new();
    for w in wires {
        xs.push(w.start.x);
        xs.push(w.end.x);
        ys.push(w.start.y);
        ys.push(w.end.y);
    }
    let mut area = (xs.iter().cloned().fold(f64::INFINITY, f64::min)
        - xs.iter().cloned().fold(f64::NEG_INFINITY, f64::max))
        .abs()
        * (ys.iter().cloned().fold(f64::INFINITY, f64::min)
            - ys.iter().cloned().fold(f64::NEG_INFINITY, f64::max))
            .abs();
    for z in keepouts {
        match z {
            KeepoutZone::Rect(r) => area -= (r.xmax - r.xmin) * (r.ymax - r.ymin),
            KeepoutZone::Circle(c) => area -= std::f64::consts::PI * c.radius * c.radius,
        }
    }
    area.max(0.0)
}

/// 对未布通线累加 penalty（增量修复用）。返回新的 LoadedData。
pub fn apply_feedback_penalties(data: &LoadedData, unrouted: &HashSet<String>, amount: f64) -> LoadedData {
    let mut wires: Vec<Wire> = data
        .wires
        .iter()
        .map(|w| {
            let mut w = w.clone();
            if unrouted.contains(&w.wire_id) {
                w.penalty += amount;
            }
            w
        })
        .collect();
    wires.shrink_to_fit();
    let mut d = data.clone();
    d.wires = wires;
    d
}

fn net_id_str(n: &Net) -> &str {
    &n.net_id
}

/// 单轮分层。`prog.set(stage,pct,msg)` 报进度，`prog.check_cancel()` 取消。
pub fn run_once(
    data: &LoadedData,
    cfg: &LayeringConfig,
    prog: &Progress,
) -> Result<LayeringResult, LayeringCancelled> {
    prog.set("分离电源/地", 3.0, "分离电源/地");
    let (trace_nets, plane_nets) = lstack::split_trace_plane(&data.nets, cfg);
    let trace_net_ids: HashSet<String> =
        trace_nets.iter().map(|n| n.net_id.clone()).collect();
    let wires: Vec<Wire> = data
        .wires
        .iter()
        .filter(|w| trace_net_ids.contains(&w.net_id))
        .cloned()
        .collect();

    let group_of: HashMap<String, String> = trace_nets
        .iter()
        .map(|n| (n.net_id.clone(), n.signal_group_id.clone().unwrap_or_else(|| "default".to_string())))
        .collect();
    let wire_group_of: HashMap<String, String> = wires
        .iter()
        .map(|w| {
            (
                w.wire_id.clone(),
                group_of.get(&w.net_id).cloned().unwrap_or_else(|| "default".to_string()),
            )
        })
        .collect();
    let allowed: HashMap<String, HashSet<i64>> = wires
        .iter()
        .map(|w| {
            (
                w.wire_id.clone(),
                lstack::wire_allowed_layers(w, &trace_nets, &data.signal_groups, data.stack.as_ref()),
            )
        })
        .collect();
    let pins: Vec<Pin> = trace_nets.iter().flat_map(|n| n.pins.clone()).collect();
    prog.log_info(&format!(
        "[阶段 分离电源/地] trace={} plane={} wires={} 允许层规则={} 引脚={}",
        trace_nets.len(),
        plane_nets.len(),
        wires.len(),
        allowed.len(),
        pins.len()
    ));

    prog.check_cancel()?;
    prog.set("构建拥塞图", 12.0, "构建拥塞图");
    let cmap0 = congestion::build_congestion_map(&wires, &data.keepouts, &pins, cfg);
    prog.check_cancel()?;
    prog.set("冲突检测", 22.0, "冲突检测");
    let (conflicts, hard_graph) = cc::detect_all_conflicts(&wires, &data.keepouts, cfg, Some(&cmap0));
    prog.log_info(&format!(
        "[阶段 冲突检测] 候选线对={} 硬冲突={} 软冲突={}",
        conflicts.len(),
        conflicts.iter().filter(|c| c.level == ConflictLevel::Hard).count(),
        conflicts.iter().filter(|c| c.level == ConflictLevel::Soft).count(),
    ));
    prog.check_cancel()?;

    let usable = _usable_area(&wires, &data.keepouts);
    let lb = layer_packing::capacity_lower_bound(&wires, &wire_group_of, usable, cfg);
    let mut warnings: Vec<String> = data.warnings.clone();

    let mut assignment: HashMap<String, i64>;
    if cfg.method == "dsatur" {
        assignment = match coloring::dsatur_color(&hard_graph, None, Some(&allowed), None) {
            Ok(a) => a,
            Err(e) => {
                warnings.push(e.0);
                HashMap::new()
            }
        };
    } else {
        prog.set("扇区轮询分层", 32.0, "扇区轮询分层");
        let mut wrap = |f: f64| {
            prog.set("扇区轮询分层", 32.0 + f * 18.0, "扇区轮询分层");
        };
        assignment = layer_packing::pack_layers(
            &wires,
            &allowed,
            &data.keepouts,
            &pins,
            cfg,
            &hard_graph,
            data.stack.as_ref(),
            Some(&mut wrap),
            prog.cancel_flag(),
        )?;
        prog.log_info(&format!(
            "[阶段 扇区轮询分层] 已分配={}/{} 线",
            assignment.len(),
            wires.len()
        ));
        let soft_pairs: Vec<(String, String)> = conflicts
            .iter()
            .filter(|c| c.level == ConflictLevel::Soft)
            .map(|c| (c.wire_a.clone(), c.wire_b.clone()))
            .collect();
        let mut layers: Vec<i64> = assignment.values().copied().collect();
        layers.sort();
        layers.dedup();
        if cfg.optimizer != "none" {
            prog.set("贪心交叉最小化", 52.0, "贪心交叉最小化");
            assignment = layer_packing::minimize_crossings(
                &assignment,
                &soft_pairs,
                &layers,
                &hard_graph,
                cfg.minimize_crossings_passes,
                prog.cancel_flag(),
            )?;
            prog.log_info("[阶段 贪心交叉最小化] 完成");
            prog.check_cancel()?;
        }
        if cfg.optimizer == "sa" {
            let mut best = assignment;
            let mut best_soft = metrics::soft_crossings(&best, &soft_pairs);
            for r in 0..(cfg.sa_restarts.max(1)) {
                let mut c = cfg.clone();
                c.sa_seed = cfg.sa_seed + r as u64;
                let mut wrap = |f: f64| {
                    prog.set("模拟退火精修", 55.0 + f * 33.0, "模拟退火精修");
                };
                let cand = optimizer::optimize_layering(
                    &best,
                    &wires,
                    &soft_pairs,
                    &hard_graph,
                    &c,
                    &allowed,
                    Some(&mut wrap),
                    prog.cancel_flag(),
                )?;
                prog.check_cancel()?;
                let s = metrics::soft_crossings(&cand, &soft_pairs);
                if s < best_soft {
                    best = cand;
                    best_soft = s;
                }
            }
            assignment = best;
            prog.log_info(&format!("[阶段 模拟退火精修] 完成 best_soft={best_soft}"));
        }
    }

    // 后处理拥塞均衡（默认关）：把超容层/格点上的线平衡到低拥塞允许层，摊平层占用峰值
    if cfg.congestion_balance {
        prog.log_info("[阶段 后处理拥塞均衡] 开始");
        let soft_pairs: Vec<(String, String)> = conflicts
            .iter()
            .filter(|c| c.level == ConflictLevel::Soft)
            .map(|c| (c.wire_a.clone(), c.wire_b.clone()))
            .collect();
        pp::congestion_balance(
            &mut assignment,
            &wires,
            &data.keepouts,
            &pins,
            &hard_graph,
            &allowed,
            &soft_pairs,
            cfg,
            prog.cancel_flag(),
        )?;
    }

    prog.check_cancel()?;
    prog.set("后处理与人工兜底", 90.0, "后处理与人工兜底");
    let viol = pp::verify_hard_free(&assignment, &hard_graph);
    let mut manual_nets: Vec<String> = Vec::new();
    let mut manual_wires: HashSet<String> = HashSet::new();
    if !viol.is_empty() {
        for (a, b) in &viol {
            manual_wires.insert(a.clone());
            manual_wires.insert(b.clone());
        }
        let mut netset: HashSet<String> = wires
            .iter()
            .filter(|w| manual_wires.contains(&w.wire_id))
            .map(|w| w.net_id.clone())
            .collect();
        let mut mn: Vec<String> = netset.drain().collect();
        mn.sort();
        manual_nets = mn;
        for w in &manual_wires {
            assignment.remove(w);
        }
        warnings.push(format!(
            "{} 条线需人工 route（同层硬冲突无法自动分层）: {:?}",
            manual_nets.len(),
            manual_nets
        ));
    }
    prog.log_info(&format!(
        "[阶段 后处理与人工兜底] 需人工={} 线",
        manual_nets.len()
    ));

    let soft_per_layer = pp::soft_conflicts_per_layer(&assignment, &conflicts);
    let (detour, via) = pp::collect_layer_marks(&wires, &assignment, &conflicts, &data.keepouts, cfg);
    let occ_per_layer = pp::max_occupancy_per_layer(&assignment, &wires, &data.keepouts, &pins, cfg);
    let (routable_net_count, total_net_count, unroutable_nets) =
        pp::routable_nets(&assignment, &wires, &data.keepouts, &pins, cfg);
    let (multi_layer_nets, via_estimate) = pp::net_span_stats(&assignment, &wires);
    // 里程碑 1：走通率的"模拟路由路径"版（按层 preferred_dir 生成 L/Z 路径）
    let mut layer_dir: HashMap<i64, String> = HashMap::new();
    if let Some(s) = data.stack.as_ref() {
        for l in &s.layers {
            layer_dir.insert(l.index, l.preferred_dir.clone());
        }
    }
    let (routable_path_net_count, _, unroutable_nets_path) =
        pp::routable_nets_path(&assignment, &wires, &data.keepouts, &pins, cfg, &layer_dir);
    // 走通率（真实可布版）：层内"容量内可走"连通区洪泛——比直线/路径占用判定更诚实（可绕行）。
    let (routable_flood_net_count, _, unroutable_nets_flood) =
        pp::routable_nets_flood(&assignment, &wires, &data.keepouts, &pins, cfg);

    let net_by_id: HashMap<String, Net> = data
        .nets
        .iter()
        .map(|n| (n.net_id.clone(), n.clone()))
        .collect();

    let mut by_layer: HashMap<i64, Vec<Wire>> = HashMap::new();
    for w in &wires {
        if let Some(&l) = assignment.get(&w.wire_id) {
            by_layer.entry(l).or_default().push(w.clone());
        }
    }
    let mut layer_keys: Vec<i64> = by_layer.keys().copied().collect();
    layer_keys.sort();

    let mut layer_infos: Vec<LayerInfo> = Vec::new();
    for layer in layer_keys {
        let ws = &by_layer[&layer];
        let mut nets: Vec<String> = ws.iter().map(|w| w.net_id.clone()).collect();
        nets.sort();
        nets.dedup();
        let mut groups: Vec<String> = nets
            .iter()
            .filter_map(|n| net_by_id.get(n).and_then(|net| net.signal_group_id.clone()))
            .collect();
        groups.sort();
        groups.dedup();
        let mut wires_ids: Vec<String> = ws.iter().map(|w| w.wire_id.clone()).collect();
        wires_ids.sort();
        let mut detour_list: Vec<String> = detour
            .get(&layer)
            .map(|s| s.iter().cloned().collect())
            .unwrap_or_default();
        detour_list.sort();
        let mut via_list: Vec<String> = via
            .get(&layer)
            .map(|s| s.iter().cloned().collect())
            .unwrap_or_default();
        via_list.sort();
        layer_infos.push(LayerInfo {
            layer_index: layer,
            kind: data
                .stack
                .as_ref()
                .map(|s| s.kind_of(layer))
                .unwrap_or_else(|| "signal".to_string()),
            signal_groups: groups,
            wires: wires_ids,
            nets,
            soft_conflict_count: soft_per_layer.get(&layer).copied().unwrap_or(0),
            max_occupancy: round4(occ_per_layer.get(&layer).copied().unwrap_or(0.0)),
            requires_detour: detour_list,
            requires_endpoint_via: via_list,
        });
    }

    if !plane_nets.is_empty() && cfg.plane_nets_excluded && data.stack.is_some() {
        let mut plane_net_ids: Vec<String> = plane_nets.iter().map(net_id_str).map(str::to_string).collect();
        plane_net_ids.sort();
        for pl in data.stack.as_ref().unwrap().plane_layers() {
            layer_infos.push(LayerInfo {
                layer_index: pl,
                kind: "plane".to_string(),
                signal_groups: Vec::new(),
                wires: Vec::new(),
                nets: plane_net_ids.clone(),
                soft_conflict_count: 0,
                max_occupancy: 0.0,
                requires_detour: Vec::new(),
                requires_endpoint_via: Vec::new(),
            });
        }
    }
    layer_infos.sort_by_key(|li| li.layer_index);

    let unassigned: Vec<String> = wires
        .iter()
        .filter(|w| !assignment.contains_key(&w.wire_id) && !manual_wires.contains(&w.wire_id))
        .map(|w| w.wire_id.clone())
        .collect();
    if !unassigned.is_empty() {
        warnings.push(format!("以下线未能在允许层内分配: {:?}", unassigned));
    }

    let hard: Vec<Conflict> = conflicts
        .iter()
        .filter(|c| c.level == ConflictLevel::Hard)
        .cloned()
        .collect();
    let soft: Vec<Conflict> = conflicts
        .iter()
        .filter(|c| c.level == ConflictLevel::Soft)
        .cloned()
        .collect();
    let mut plane_nets_ids: Vec<String> = plane_nets.iter().map(net_id_str).map(str::to_string).collect();
    plane_nets_ids.sort();

    prog.set("完成", 100.0, "完成");
    Ok(LayeringResult {
        layers: layer_infos,
        assignment,
        plane_nets: plane_nets_ids,
        hard_conflicts: hard,
        soft_conflicts: soft,
        method: cfg.method.clone(),
        iterations_used: 1,
        capacity_lower_bound: lb,
        warnings,
        manual_route_nets: manual_nets,
        routable_net_count,
        total_net_count,
        unroutable_nets,
        routable_path_net_count,
        unroutable_nets_path,
        routable_flood_net_count,
        unroutable_nets_flood,
        multi_layer_nets,
        via_estimate,
    })
}

fn round4(x: f64) -> f64 {
    (x * 10000.0).round() / 10000.0
}

/// 单轮分层；反馈闭环为占位（插件当前不传 feedback_path，等价单轮）。
pub fn run(
    data: &LoadedData,
    cfg: &LayeringConfig,
    feedback_path: Option<&str>,
    prog: &Progress,
) -> Result<LayeringResult, LayeringCancelled> {
    let _ = feedback_path;
    run_once(data, cfg, prog)
}
